packages/vite/src/node/server/hmr.ts

```diff
@@ -116,14 +170,176 @@ export interface HotChannel {
+    | undefined
+
+  return {
+    ...channel,
+    on: (
+      event: string,
+      fn: (data: any, client: NormalizedHotChannelClient | undefined) => void,
+    ) => {
+      if (event === 'connection') {
+        channel.on?.(event, fn as () => void)
+        return
+      }
+
+      const listenerWithNormalizedClient = (
+        data: any,
+        client: HotChannelClient | undefined,
+      ) => {
+        if (!client) {
+          fn(data, undefined)
+          return
+        }
+        const normalizedClient: NormalizedHotChannelClient = {
+          send: (...args) => {
+            let payload: HotPayload
+            if (typeof args[0] === 'string') {
+              payload = {
+                type: 'custom',
+                event: args[0],
+                data: args[1],
+              }
+            } else {
+              payload = args[0]
+            }
+            client.send(payload)
+          },
+        }
+        fn(data, normalizedClient)
+      }
+      normalizedListenerMap.set(fn, listenerWithNormalizedClient)
+
+      channel.on?.(event, listenerWithNormalizedClient)
+      if (!listenersForEvents.has(event)) {
+        listenersForEvents.set(event, new Set())
+      }
+      listenersForEvents.get(event)!.add(listenerWithNormalizedClient)
+    },
+    off: (event: string, fn: () => void) => {
+      if (event === 'connection') {
+        channel.off?.(event, fn as () => void)
+        return
+      }
+
+      const normalizedListener = normalizedListenerMap.get(fn)
+      if (normalizedListener) {
+        channel.off?.(event, normalizedListener)
+        listenersForEvents.get(event)?.delete(normalizedListener)
+      }
+    },
+    setInvokeHandler(invokeHandlers) {
+      if (!invokeHandlers) {
+        channel.setInvokeHandler?.(invokeHandlers)
+        if (listenerForInvokeHandler) {
+          channel.off?.('vite:invoke', listenerForInvokeHandler)
+        }
+        return
+      }
+
+      const wrappedInvokeHandler = async <T extends keyof InvokeMethods>(
+        payload: HotPayload,
+      ) => {
+        if (payload.type !== 'custom') return
+        if (payload.event === 'vite:invoke') {
+          const data: InvokeSendData<T> = payload.data
+          const { name, data: args } = data
+          try {
+            // @ts-expect-error `invokeHandlers[name]` is `InvokeMethods[T]`, so passing the args is fine
+            const result = await invokeHandlers[name](...args)
+            return { r: result }
+          } catch (error) {
+            return {
+              e: {
+                name: error.name,
+                message: error.message,
+                stack: error.stack,
+              },
+            }
+          }
+        } else {
+          const listeners = listenersForEvents.get(payload.event)
+          if (listeners) {
+            for (const listener of listeners) {
+              listener(payload.data, undefined)
+            }
+          }
```

Comment:
I forgot about thinking about this one. When `invoke` is used as `send`, we cannot make `client` that can be passed to the listeners. Making it possible to pass in `client` from the user side makes the API strange (`const value = invokeHandler(payload, client)`).

We can make a `client` that only supports sending a value when the listener is not finished (i.e. make listener to return a promise and collect the values passed to `client.send` until that promise is resolved). But this feels like we are making multiple ways to do the same thing with `invoke`.

We can leave this as this implementation, but that makes the listener's client to be optional, and makes the API a bit difficult to use.
```js
env.hot.on('foo', (data, client) => {
  if (client) {
    client.send('bar')
  } else {
    // frameworks: what should I do here?
  }
})
```

Maybe we should just remove `const sendOrInvoke = transport.send ?? transport.invoke` functionality entirely.

Reply:
I vote for removing `sendOrInvoke` . My understanding is that the consequence of it is only preventing `invoke`-only transport from using `import.meta.hot.send` on runner, but `import.meta.hot` API wouldn't be fully available anyways in that case. Is my understanding correct?

Reply:
Yeah, makes sense. Removed in d55aec86adf06ce4e42639829ce9e3e50692966f .